API/ev3_timer.c

```diff
@@ -12,268 +16,552 @@
+    RETURN_ON_FAIL(result, "read(timerfd) failed", false);
+
+    // and finally, call the function the needed number of times
+    for (uint64_t n = 0; n < elapsedExpirations; n++)
+        tmArrayCallAll(pLocalCopy, interval);
+
+    return true;
+}
+
+
+void tmArrayCallAll(CallbackBlock *pBlock, TimerInterval interval)
+{
+    // prevent SIGSEGV
+    if (pBlock->elements == NULL)
+        return;
+
+    // call all the functions
+    for (size_t i = 0; i < pBlock->count; i++)
+        pBlock->elements[i](interval);
+}
+
+// DESTRUCTION //
+
+bool TimerExit()
 {
-  if (Timer < NUM_MS_TIMERS)
-	return TimerGetMS() - msTimers[Timer];
-  return 0;
+    bool success = true;
+
+    if (!TimerInitialized())
+        return true;
+
+    // stop the waiter thread
+    success &= tmExitThread();
+    // clean up kernel objects
+    success &= tmExitKernel();
+    // free callback array allocations
+    tmFreeArrays();
+
+    tcb.initialized = false;
+    return success;
 }
 
-void ClearTimerMS(uint8_t Timer)
+bool tmExitThread(void)
 {
-  if (Timer < NUM_MS_TIMERS)
-	msTimers[Timer] = TimerGetMS();
+    bool success = true;
+
+    // notify the thread that it should quit
+    SET_ON_FAIL(eventfd_write(tcb.waiterTerminatorFd, 1),
+                "eventfd_write() failed", success, false);
+
+    // wait for the thread to terminate
+    SET_ON_FAIL(pthread_join(tcb.waiterThread, NULL),
+                "pthread_join() failed", success, false);
+
+    // and clean up the waiter mutex
+    SET_ON_FAIL(pthread_mutex_destroy(&tcb.waiterMutex),
+                "pthread_mutex_destroy(waiter) failed", success, false);
+
+    // and the timestamp mutex
+    SET_ON_FAIL(pthread_mutex_destroy(&tcb.timestampMutex),
+                "pthread_mutex_destroy(timestamp) failed", success, false);
+
+    return success;
 }
 
-void SetTimerMS(uint8_t Timer, unsigned long Value)
+bool tmExitKernel()
 {
-  if (Timer < NUM_MS_TIMERS)
-	msTimers[Timer] = TimerGetMS() - Value;
+    bool success = true;
+
+    for (TimerInterval i = ti10ms; i < ti_count; i++)
+    {
+        // close timer objects
+        int fd = tcb.callbacksFor[i].timer_fd;
+        if (fd != -1)
+            SET_ON_FAIL(close(fd), "close(timerfd) failed", success, false);
+
+        tcb.callbacksFor[i].timer_fd = -1;
+    }
+
+    // close the eventfd terminator
+    if (tcb.waiterTerminatorFd != -1)
+        SET_ON_FAIL(close(tcb.waiterTerminatorFd), "close(eventfd) failed", success, false);
+
+    // and close the epoll waiter as well
+    if (tcb.waiterEpollFd != -1)
+        SET_ON_FAIL(close(tcb.waiterEpollFd), "close(epollfd) failed", success, false);
+
+    tcb.waiterTerminatorFd = -1;
+    tcb.waiterEpollFd = -1;
+    return success;
 }
 
-unsigned long TimerUS(uint8_t Timer)
+void tmFreeArrays()
 {
-  if (Timer < NUM_US_TIMERS)
-	return TimerGetUS() - usTimers[Timer];
-  return 0;
+    for (TimerInterval i = ti10ms; i < ti_count; i++)
+        tmArrayFree(&tcb.callbacksFor[i]);

```

Comment:
racy: you can free the array, while tmArrayMirror is running

Reply:
This is partially an unsolved issue in this PR. During the runtime (after `TimerInit()`), the module's functions should be thread safe. However `TimerInit()` and `TimerExit()` themselves are not necessarily thread safe. The `tmFreeArrays()` function is called only a a last step of `TimerExit()`. At that time, the thread *should* be stopped (it is joined beforehand, although that could fail). Therefore the module hopefully isn't racy with itself, but if something external is mirroring an array during the `TimerExit()` call, that will cause a race condition.

During the runtime, the top-level functions lock the two global mutexes that prevent a race condition from happening. These lower-level functions then rely on their caller to make sure that a race condition will not happen. During the call to `TimerExit()`, the two global mutexes don't exist anymore, so an alternative mechanism has to be devised.

Reply:
I think that adding a mutex and a initialization flag to the CallbackArray could solve this. This could be useful if the resizable array implementation is promoted to a public API in the future.

Reply:
This should now be mostly solved by introducing a statically-initialized mutex and protecting all critical code with it.